pkg/prober/prober.go

```diff
@@ -0,0 +1,160 @@
+package prober
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/component"
+	"github.com/prometheus/common/route"
+)
+
+const (
+	healthyEndpointPath  = "/-/healthy"
+	readyEndpointPath    = "/-/ready"
+	okProbeFmt           = "thanos %v is %v"
+	errorProbeFmt        = "thanos %v is not %v. Reason: %v"
+	probeErrorHTTPStatus = 503
+	initialErrorFmt      = "thanos %s is initializing"
+)
+
+// Prober represents health and readiness status of given component.
+type Prober struct {
+	logger             log.Logger
+	componentMtx       sync.RWMutex
+	component          component.Component
+	readyMtx           sync.RWMutex
+	readiness          error
+	healthyMtx         sync.RWMutex
+	healthiness        error
+	readyStateMetric   prometheus.Gauge
+	healthyStateMetric prometheus.Gauge
+}
+
+// NewProber returns Prober representing readiness and healthiness of given component.
+func NewProber(component component.Component, logger log.Logger, reg prometheus.Registerer) *Prober {
+	initialErr := fmt.Errorf(initialErrorFmt, component)
+
+	p := &Prober{
+		component:   component,
+		logger:      logger,
+		healthiness: initialErr,
+		readiness:   initialErr,
+		readyStateMetric: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: fmt.Sprintf("thanos_%s_ready", component),
+			Help: "Represents readiness status of the thanos component.",
+		}),
+		healthyStateMetric: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: fmt.Sprintf("thanos_%s_healthy", component),
+			Help: "Represents healthiness status of the thanos component.",
+		}),
+	}
+	if reg != nil {
+		reg.MustRegister(p.readyStateMetric, p.healthyStateMetric)
+	}
+	return p
+}
+
+// RegisterInRouter registers readiness and liveness probes to router.
+func (p *Prober) RegisterInRouter(router *route.Router) {
+	router.Get(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	router.Get(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+// RegisterInMux registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInMux(mux *http.ServeMux) {
+	mux.HandleFunc(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	mux.HandleFunc(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+func (p *Prober) writeResponse(w http.ResponseWriter, probeFunc func() error, probeType string) {
+	err := probeFunc()
```

Comment:
you can do `if err := probeFn(); err != nil ` here (: Also we usually do `Fn` suffix for function vars so  `probeFn`  would be nice? (: What do you think?